SOLR-15367 Convert "rid" functionality into a default Tracer

stillalex opened the pull request:
https://issues.apache.org/jira/browse/SOLR-15367

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

Trace Id propagation only - based on OTEL mechanism. this will only pickup the id from the incoming request (or generate a new one) and make sure to attach it to all internal communication.
It can be a much better alternative to the existing rid mechanism, without necessarily replacing it.

Can be enabled via system property (under evaluation if it can be enabled by default). this is located in solr core and works with the existing otel libraries and always-on tracing mechanism.

The header is configurable, and could be extended into a list of headers. I would be ok with seeding this from a system property to make config simpler.


# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

stillalex commented:
@dsmiley would love your thoughts on this one.

dsmiley reviewed (commented):
Looks good to me!

In a separate issue we could remove the existing "rid".

It's kind of a shame to require the whole open tracing SDK for this functionality but so be it.

stillalex commented on solr/modules/opentelemetry/src/test/org/apache/solr/opentelemetry/TestSimplePropagatorDistributedTracing.java:
```diff
@@ -0,0 +1,101 @@
+    }
+  }
+
+  private void assertSameTraceId(LogListener reqLog) {
```
yeah, good catch

stillalex commented on solr/modules/opentelemetry/src/test/org/apache/solr/opentelemetry/TestSimplePropagatorDistributedTracing.java:
```diff
@@ -0,0 +1,101 @@
+    // TODO check tid
+
+    try (LogListener reqLog = LogListener.info(SolrCore.class.getName() + ".Request")) {
+      assertNull(MDC.get(MDCLoggingContext.TRACE_ID));
```
the check is to verify is the context data is actually null. maybe the placement for the check is not ideal.

dsmiley replied:
From the context of the test, it might always be null?  Maybe you intended to check this from within Solr (thus during a request).  I suppose what you should be looking for is the "t" MDC value in the logs.
I think I'm not understanding something.  Don't we want to actually see a trace ID after all?  Even for "only" propagation; nonetheless there is a trace ID and logging it is absolutely the point.

stillalex replied:
I think this was distracting from the main conversation. I can say it was my mistake for copy/pasta from the TestRequestId test and I removed it from the updated test. we don't need to get into it further. for the rest I agree with checking the logging context for trace ids as the main function of the test.

stillalex commented on solr/modules/opentelemetry/src/test/org/apache/solr/opentelemetry/TestSimplePropagatorDistributedTracing.java:
```diff
@@ -0,0 +1,101 @@
+
+    // tracer should be disabled
+    assertEquals(
+        "Expecting active otel, not noop impl",
```
yeah, error message is wrong, will fix it

stillalex commented on solr/modules/opentelemetry/src/java/org/apache/solr/opentelemetry/simplepropagator/SimplePropagator.java:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  private static String newTraceId() {
+    return String.format("%s-%d", UUID.randomUUID(), traceCounter.incrementAndGet());
```
ok. makes sense will keep the existing mechanism.

stillalex commented:
> It's kind of a shame to require the whole open tracing SDK for this functionality but so be it.

you raise a very interesting point, and I think this could work without the sdk. it only really needs a barebones otel impl to make this work, and everything should already be in solr-core (otel api and otel context). if it does turn out it's light on dependencies this could actually move to solr core and no longer depend on the otel module to function. let me check this.

stillalex commented:
@dsmiley thank you for the review so far. I managed to successfully move the functionality to solr-core without extra dependencies (just api and context). I think this allows for much more freedom with enabling by default and eventually (maybe) replacing the rid.
I updated the test and something weird failed. it seems the `UpdateRequest` does not actually propagate headers to internal communication? I'm not even sure if this is by design or not. I added a todo to verify.

My plan (given this moved to core) would be to turn this into a 'on by default' behavior (add a flag to allow disabling). Optionally I would source rid from this value, so that there is no confusion between multiple similar sets of values.
Please take another look and let me know what you think so far.

stillalex commented on solr/core/src/test/org/apache/solr/util/tracing/TestSimplePropagatorDistributedTracing.java:
```diff
@@ -0,0 +1,130 @@
+      u.addHeader(SimplePropagator.TRACE_ID, traceId);
+      var r1 = u.process(cloudClient, COLLECTION);
+      assertEquals(0, r1.getStatus());
+      assertSameTraceId(reqLog, traceId);
```
this only works if `UpdateRequest` enables header propagation, see change below.

dsmiley replied:
weird.

dsmiley commented:
I'm so glad you managed to implement this in core!  Changes looks good.

FYI, one gap in propagation is that our Apache HttpClient isn't instrumented with tracing, only the Jetty HttpClient.  I actually have a fix for this locally.  Of course, ideally we would only use Jetty HttpClient.

stillalex commented:
> FYI, one gap in propagation is that our Apache HttpClient isn't instrumented with tracing, only the Jetty HttpClient. I actually have a fix for this locally.

could you be a bit more specific? I am not sure where the gap is, the context injection (addition of http headers) should ideally happen on both client, but maybe you have seen a specific flow that is broken.

I am down a rabbit hole now looking for `new UpdateRequest()` calls that wipe out the headers so propagation would be broken.

dsmiley commented:
My local change is to `InstrumentedHttpRequestExecutor` to add a client span and to inject trace headers.  This class is used in UpdateShardHandler to create the "defaultClient" (Apache HttpClient).  Without this, any communication using Apache HttpClient won't propagate tracing.  If you were to make changes about passing certain headers in a generic way, that would be a work-around but the ideal fix is one I made and could share in InstrumentedHttpRequestExecutor.

stillalex commented:
I think I understand what you are saying. but I think it was misleading to say propagation doesn't work (I think you partially corrected later by saying "won't propagate tracing"). a more correct thing to say is 'tracing in general does not work with the apache client' and so this propagation proposal is broken due to the same underlying cause. the fix would be for the otel tracing in general and it would also enable the propagation. along the same lines, the existing 9.x tracing impl is broken due to the same reason I believe, the apache client is not aware of the trace context. please correct if I am misunderstanding your point.
I see why you would choose `InstrumentedHttpRequestExecutor`. it offers a cross cutting access point into the execution, but I don't think it's a very good idea to mix functionality like this (metrics collection with tracing headers). I am looking at alternatives, maybe we could setup a `HttpRequestInterceptor` which would allow for a cleaner separation of concerns.

dsmiley commented:
> I don't think it's a very good idea to mix functionality like this (metrics collection with tracing headers)

I disagree; both are "instrumentation" and the class name was already chosen in this generic way instead of being called MetricsBlahBlahBlah.  Nonetheless the concerns could be split up to work in parallel but it's extra code for unclear gain.

BTW again in my fork locally, I did similarly for the Jetty side in InstrumentedHttpListenerFactory moving the previous logic to this central location, and I think in so doing picked up more use-cases.

stillalex commented:
> I disagree; both are "instrumentation" and the class name was already chosen in this generic way instead of being called MetricsBlahBlahBlah. Nonetheless the concerns could be split up to work in parallel but it's extra code for unclear gain.

sure, I see the point and I don't disagree with it.

> BTW again in my fork locally, I did similarly for the Jetty side in InstrumentedHttpListenerFactory moving the previous logic to this central location, and I think in so doing picked up more use-cases.

is this code you mention open source? if so could you link to it? it would make the discussion much easier.

stillalex commented:
I took a stab at having this on by default,  just to see what the tests say. and it looks ok so far. I am also open to not doing this right away, to allow for some more testing time.

dsmiley reviewed (commented):
It's worth noting that the new propagator bears resemblance to io.opentelemetry.extension.incubator.propagation.PassThroughPropagator -- should probably reference this fact in the javadocs.

An interesting thing I noticed when browsing OTEL is the javadocs of Context refer to `-Dio.opentelemetry.context.enableStrictContext=true` as being something advisable to set for tests for enhanced checking that Scopes, etc. are closed properly.  It's trivial enough that you could just do it in this PR.

stillalex commented on solr/core/src/java/org/apache/solr/util/stats/InstrumentedHttpRequestExecutor.java:
```diff
@@ -147,6 +152,13 @@ public HttpResponse execute(HttpRequest request, HttpClientConnection conn, Http
     }
   }
 
+  private static final TextMapSetter<HttpRequest> TRACE_INJECTOR = (r, k, v) -> r.setHeader(k, v);
+
+  private static void injectContextIntoRequest(HttpRequest request) {
+    TextMapPropagator textMapPropagator = TraceUtils.getTextMapPropagator();
+    textMapPropagator.inject(Context.current(), request, TRACE_INJECTOR);
+  }
+
```
> Is declaring TRACE_INJECTOR useful for a performance perspective? It kinda makes me sad to not see a trivial lambda used as such.

yeah it's coming from their javadocs (both TextMapSetter and TextMapGetter)
```
 * <p>{@code Setter} is stateless and allows to be saved as a constant to avoid runtime allocations.
```

stillalex replied:
> Worthy of it's own JIRA issue IMO.

agreed. I wanted to cover all feedback received on the PR and in doing so, I lost focus a bit. I will break this off into its own Jira.

stillalex commented on solr/core/src/test/org/apache/solr/util/tracing/TestSimplePropagatorDistributedTracing.java:
```diff
@@ -114,6 +115,26 @@ public void test() throws IOException, SolrServerException {
 
+  @Test
+  public void testApacheClient() throws Exception {
+    // call api that uses apache client impl
```
sure I even mentioned this in my own PR comment

> the test is fragile, as it assumes the api call will rely on apache http client impl, if this were to change in the future the test would not be useful anymore.

stillalex commented:
@dsmiley did a recap above and pushed a few more commits. I would like to leave it on by default to catch regressions early, I will disable it if it turns out to be too noisy. 
I think this is good to go, please let me know if you see anything blocking.

dsmiley commented on solr/test-framework/src/java/org/apache/solr/cloud/MiniSolrCloudCluster.java:
```diff
@@ -1237,6 +1240,11 @@ public MiniSolrCloudCluster build() throws Exception {
           "solr.distributedClusterStateUpdates",
           Boolean.toString(useDistributedClusterStateUpdate));
 
+      // eager init to prevent OTEL init races caused by test setup
+      if (!disableTraceIdGeneration && TracerConfigurator.TRACE_ID_GEN_ENABLED) {
+        SimplePropagator.load();
+      }
+
```
This part seems suspicious because you are adding this to MiniSolrCloudCluster.  Tracing isn't SolrCloud specific and isn't limited to this utility.  Maybe this should be in STCJ4 as there used to be something which you removed (not sure why)?

stillalex replied:
I removed something that was not needed anymore https://github.com/apache/solr/pull/1841/files#diff-7cb8f3a46dbdb99ccfdcc1d07c4457167188e7366534a7e7e4a0b5aff93c1cc4L327

this is here because it's the place that boots up a multi node cluster and causing multiple OTEL inits to happen in the same jvm. I believe this is the good place for it because this is where the cluster gets setup. it is not suited for the STCJ4.

stillalex replied:
just checking I read your message correctly. are you starting multiple solr instances as a part of a test and not using the MiniSolrCloudCluster? in this case, sure I can consider moving so it covers all solr flavors. if it's not a multi solr setup, the race does not happen and there is no issue.

dsmiley replied:
Oh I see.  Well there are number of tests manually working with multiple JettySolrRunner instances but they don't count?  Is it because we need only concern ourselves with tests touching tracing?

stillalex replied:
> Is it because we need only concern ourselves with tests touching tracing?

no, with the "always on trace id" all tests are using tracing.

the core issue is with OTEL init method. it will not allow setting the global object more than once and will fail loudly if you try, so we'd see a test failure. I am assuming the other tests do not have the same level of concurrency on multiple solr instances coming online, so maybe the race is less likely to occur as in the case of the MiniSolrCloudCluster.

dsmiley replied:
Oh this is a race... initing OTEL _concurrently_ ?  Then I agree MiniSolrCloudCluster is probably the only way this would happen.  I thought you meant merely multiple times (e.g. in sequence).

Any way, a code comment explaining this more would help but regardless I'm +1
